leshan-core/src/main/java/org/eclipse/leshan/core/node/TimestampedLwM2mNodes.java

```diff
@@ -0,0 +1,164 @@
+            result.putAll(entry.getValue());
+        }
+        return result;
+    }
+
+    /**
+     * Returns the all sorted timestamps of contained nodes with ascending order. Null timestamp is considered as most
+     * recent one.
+     */
+    public Set<Long> getTimestamps() {
+        return timestampedPathNodesMap.keySet();
+    }
+
+    @Override
+    public String toString() {
+        return String.format("TimestampedLwM2mNodes [timestampedPathNodesMap=%s, timestampedNodes=%s]",
+                timestampedPathNodesMap);
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + ((timestampedPathNodesMap == null) ? 0 : timestampedPathNodesMap.hashCode());
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
+        if (obj == null)
+            return false;
+        if (getClass() != obj.getClass())
+            return false;
+        TimestampedLwM2mNodes other = (TimestampedLwM2mNodes) obj;
+        if (timestampedPathNodesMap == null) {
+            if (other.timestampedPathNodesMap != null)
+                return false;
+        } else if (!timestampedPathNodesMap.equals(other.timestampedPathNodesMap))
+            return false;
+        return true;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+
+        private final Map<Long, Map<LwM2mPath, LwM2mNode>> timestampedPathNodesMap = getTimestampedMap();
+
+        public Builder addNodes(Map<LwM2mPath, LwM2mNode> pathNodesMap) {
+            timestampedPathNodesMap.put(null, pathNodesMap);
+            return this;
+        }
+
+        public Builder put(Long timestamp, LwM2mPath path, LwM2mNode node) {
+            if (!timestampedPathNodesMap.containsKey(timestamp)) {
+                timestampedPathNodesMap.put(timestamp, new HashMap<>());
+            }
+            timestampedPathNodesMap.get(timestamp).put(path, node);
+            return this;
+        }
+
+        public Builder put(LwM2mPath path, LwM2mNode node) {
+            put(null, path, node);
+            return this;
+        }
+
+        public Builder add(TimestampedLwM2mNodes timestampedNodes) {
+            for (Map.Entry<Long, Map<LwM2mPath, LwM2mNode>> entry : timestampedNodes.getTimestampedNodes().entrySet()) {
+                Long timestamp = entry.getKey();
+                Map<LwM2mPath, LwM2mNode> pathNodeMap = entry.getValue();
+
+                for (Map.Entry<LwM2mPath, LwM2mNode> pathNodeEntry : pathNodeMap.entrySet()) {
+                    LwM2mPath path = pathNodeEntry.getKey();
+                    LwM2mNode node = pathNodeEntry.getValue();
+                    put(timestamp, path, node);
+                }
+            }
+            return this;
+        }
```

Comment:
We should maybe add some check here and raise exception in case of duplicate data ?
(this exception should be catched in decoder)

Reply:
I'm not sure I was clear.
My point is when you put new data in the `Builer.timestampedPathNodesMap` structure,  overrinding already existing data seems to be not valid to me ?

Reply:
I'm understand and I'm still think that should not be forbidden, it doesn't sound bad for me.

Reply:
I wanted to be sure before to go in deeper explanation. :slightly_smiling_face:

Let me explain why I think that this should **maybe** not be allowed.

I feel there two usage about this: 
1. timestampedNodes created in a decoder
2. timestampedNode created at client side by the library itself.

In first case, I consider if that duplicate value means payload we try to decode in invalid and so I think we should raise an exception. 

In second case, I'm not sure because we don't implement anything about this but I also bet that trying to store 2 value for same path and same timestamp looks like a bug. 

Did I missed something ? 

In case, we really want to allow to override value in timestampedNodes, maybe we can adapt the API (no more fluent builder) and return erased nodes (when duplicate) but I'm not sure if this worth it  because I mainly see use case where we don't want duplicate :thinking:

Reply:
Ok, the argument about validation in decoder is ok for me - we should not allow to decode invalid data. But my intention for TimestampedLwM2mNodes.Builder was not only decoder usage, so maybe we should allow to add validation optionally e.g. ```TimestampedLwM2mNodes.builder(boolean forbidDuplicates)``` or ```builder.forbidDuplicates(true)``` (it's only proposition).

> maybe we can adapt the API (no more fluent builder)

The reason to create builder is to make TimestampedLwM2mNodes immutable. You can technically create many constructors that allow to create TimestampedLwM2mNodes in many ways, but it's much easier using builder.

Reply:
> The reason to create builder is to make TimestampedLwM2mNodes immutable. You can technically create many constructors that allow to create TimestampedLwM2mNodes in many ways, but it's much easier using builder.

I mean we keep the builder but not more fluent anymore (as we use returned value for something else than returning the builder itself)

Reply:
I don't want to spoil builder fluent. Maybe better option would be add getter to get/check the value.

Reply:
> Maybe better option would be add getter to get/check the value

I'm not sure to see how ? could you detailed this ?

Reply:
(I just try to work on this, I will propose a solution soon)

Reply:
See last commit about validation.
Is there something like this you have in mind ?

Reply:
I looked at builder update and the solution is great for me :+1: